lib/Targets/AIETargetAirbin.cpp

```diff
@@ -0,0 +1,1250 @@
+ * --------------------------------------------
+ * | Array offset | Column | Row | Tile addr  |
+ * --------------------------------------------
+ */
+class TileAddress {
+public:
+  TileAddress(uint8_t column, uint8_t row, uint64_t array_offset = 0x000u)
+      : array_offset{array_offset}, column{column}, row{row} {}
+
+  // SFINAE is used here to choose the copy constructor for `TileAddress`,
+  // and this constructor for all other classes.
+  template <
+      typename Op,
+      std::enable_if_t<not std::is_same<Op, TileAddress>::value, bool> = true>
+  TileAddress(Op &op)
+      : TileAddress{static_cast<uint8_t>(op.colIndex()),
+                    static_cast<uint8_t>(op.rowIndex())} {}
+
+  uint64_t fullAddress(uint32_t register_offset) const {
+    return (array_offset << TILE_ADDR_ARR_SHIFT) |
+           (static_cast<uint64_t>(column) << TILE_ADDR_COL_SHIFT) |
+           (static_cast<uint64_t>(row) << TILE_ADDR_ROW_SHIFT) |
+           register_offset;
+  }
+
+  bool isShim() const { return row == 0; }
+
+  operator uint16_t() const {
+    return (static_cast<uint16_t>(column) << TILE_ADDR_ROW_WIDTH) | row;
+  }
+
+  uint8_t col() const { return column; }
+
+  void clearRange(uint32_t range_start, uint32_t length);
+
+private:
+  uint64_t array_offset : 34;
+  uint8_t column : TILE_ADDR_COL_WIDTH;
+  uint8_t row : TILE_ADDR_ROW_WIDTH;
+};
+
+static_assert(sizeof(TileAddress) <= sizeof(uint64_t),
+              "Tile addresses are at most 64-bits");
+
+class Address {
+public:
+  Address(TileAddress tile, uint32_t offset) : tile{tile}, offset{offset} {}
+
+  operator uint64_t() const { return tile.fullAddress(offset); }
+
+  TileAddress destTile() const { return tile; }
+  uint32_t get_offset() const { return offset; }
+
+private:
+  TileAddress tile;
+  uint32_t offset : TILE_ADDR_OFF_WIDTH;
+};
+
+typedef std::pair<uint64_t, uint32_t> Write;
+
+class Section {
+public:
+  Section(uint64_t addr) : address(addr){};
+  uint64_t get_addr(void) const { return address; }
+  size_t get_length(void) const { return data.size() * sizeof(uint32_t); }
+  void add_data(uint32_t value) { data.push_back(value); }
+  const uint32_t *get_data(void) const { return data.data(); }
+
+private:
+  uint64_t address;           // start address of this section
+  std::vector<uint32_t> data; // data to be written starting at 'address'
+};
+
+// This template can be instantiated to represent a bitfield in a register.
+template <uint8_t high_bit, uint8_t low_bit = high_bit> class Field final {
+public:
+  static_assert(high_bit >= low_bit,
+                "The high bit should be higher than the low bit");
+  static_assert(high_bit < sizeof(uint32_t) * 8u,
+                "The field must live in a 32-bit register");
+
+  static constexpr auto num_bits_used = (high_bit - low_bit) + 1u;
+  static constexpr auto unshifted_mask = (1u << num_bits_used) - 1u;
+  static_assert((low_bit != high_bit) xor (unshifted_mask == 1),
```

Comment:
```suggestion
  static_assert((low_bit != high_bit) ^ (unshifted_mask == 1),
```
I think the rationale nowadays is that we have serious encoding nowadays and do not use digraph or trigraph anymore https://en.cppreference.com/w/cpp/language/operator_alternative

Reply:
done